### PR TITLE
[CMake] Add basic CMakeLists.txt file

### PR DESCRIPTION
**Pull request by Mike-Devel**
The file just processes all CMakeLists.txt files
that it findes in one of the library or tool root
directories.

Further reference: https://groups.google.com/forum/#!topic/boost-developers-archive/kM4JRmyVl3M

### CODE REVIEW

**Comment by Mike-Devel** [comment]
From the ML https://groups.google.com/forum/#!topic/boost-developers-archive/kM4JRmyVl3M (From James E. King III ):

> Absolutely, positively do NOT use GLOB for project build chains or for
> header or source discovery.
> There are well documented reasons not to do this.  You must express the
> dependency chain in
> linear order in libs/CMakeLists.txt.
> 
> I would have the boostorg/CMakeLists.txt call:
> 
> add_subdirectories ( libs )
> after it has included all the things libs depends on.
> 
> Then have libs/CMakeLists.txt call add_subdirectories in dependency order.
> That is the only way you can reasonably ensure a linear dependency chain.

So if that is what you want, I can turn the globbing into an explicit list of libraries.

However, before I do that, could soemone please explain more detailed what specifically are the problems with globbing in the particular context of this CMakeLists.txt file. I believe I know all the arguments for not globbing for source files but imho most don't really apply here.

1) CMake is perfectly capable of resolving dependencies it sees in reverse order. So I don't know, why the libraries need to be passed in dependency order (maintaining that order correctly seems to be even more work than maintaining the list of libraries that do have a CMakeLists.txt file - who would maintain it?)
2) Not globbing means that every time a new library adds a cmake file and wants 
   to hook up to the project wide mechanism, a new PR to the top level project 
   needs to be merged (in addition to any automatic update of the submodule). 
   I can see that this can be seen as an advantage as well as a disadvantage,
   but wanted to call it out regardless.
3) Even if no globbing is being used, there should be a mechanism to only add
   a subdirectory if the folder/CMakeFile it is actually present (often I don't 
   want to clone all of boost but just a defined subset). 

Regardless of globbing/non-globbing I intent to implement a blacklist that lists ignored libraries (e.g. because their CMakeLists.txt file doesn't play nice - yet).

**Comment by Mike-Devel** [comment]
> I assume the same would be true for subdirectories as well.

Imho this is a completely different situation for the reasons I explained above. Most importantly the support for partial checkout and because the person that adds the CMakeFile (the library maintainer) usually doesn't have the rights to modify this top level cmake file accordingly.

Also, if I translate above quote to the current situation it would read:

> If the top level CMakeFile doesn't change when a new library CMakeFile is added or removed then the generated build system cannot know when to ask CMake to regenerate.)

In other words: cmake needs to be called again manually when a new cmake file is added orp moved (not when on changes). How is that a problem and is the alternative really preferable?

I guess what I'm asking for is a different explanation than "I've heard globbing (in a different context) is discouraged"

**Comment by pdimov** [comment]
I'm not sure that we want to include the tools here. At some point, we want `cmake --build . --target install` to perform the equivalent of `b2 install`, and that doesn't install any tools. That's probably for the better, as their names are too generic for `bin` (except perhaps for `boostdep` :-) ).

For installing the tools, I think that we'll be fine with supporting that only from the appropriate tool's directories, one by one, without having a superproject batch tool install.

Maybe we do need to test the tools as part of `cmake --build . --target check` though. Without knowing what our testing setup is going to look like, it's hard to tell. We could accumulate all the tests at once, or we could `ctest` the libraries one by one in the `check` target.

In either case, it might be better to split that into `libs/CMakeLists.txt` and `tools/CMakeLists.txt` and only leave the two `add_subdirectory` calls here (if we even have `tools/CMakeLists.txt` at all).

I'm not particularly fond of the variables either. We should just have a hardcoded blacklist. We know which libraries don't work.

**Comment by Mike-Devel** [comment]
> I'm not sure that we want to include the tools here. At some point, we want cmake --build . --target install to perform the equivalent of b2 install, and that doesn't install any tools. That's probably for the better, as their names are too generic for bin (except perhaps for boostdep :-) ).

Well, at least some tools use boost libraries, so they could profit from the targets made available via this super-project. For now I'm just deactivating tool building by default, but if you feel strongly about this I'm fine with completely removing them

> For installing the tools, I think that we'll be fine with supporting that only from the appropriate tool's directories, one by one, without having a superproject batch tool install.

> Maybe we do need to test the tools as part of cmake --build . --target check though. Without knowing what our testing setup is going to look like, it's hard to tell. We could accumulate all the tests at once, or we could ctest the libraries one by one in the check target.

Well, installation and testing is a whole nother can of worms that will hopefully get resolved during the cmake review on the ML. I don't really expect this file to live past that review. 


> In either case, it might be better to split that into libs/CMakeLists.txt and tools/CMakeLists.txt and only leave the two add_subdirectory calls here (if we even have tools/CMakeLists.txt at all).

I can do that, but personally I don't think the file is complex enough yet to warrant the splitup (People then have to look at three files instead of one to understand what is going on).

> I'm not particularly fond of the variables either. We should just have a hardcoded blacklist. We know which libraries don't work.

Not really. For one, compatibility in individual libraries is currently a moving target. Having (only) a fixed black list would require a lot of PRs to the super project. Second, some libraries are in principle usable, but mutually exclusive (name collision). Third, compatibility also depends on what branches are currently checked out and which target & toolchain they are compiled for / with.

Of course that doesn't mean there shouldn't be a sensible (& conservative) default, but having a easy option to override this list without changing the actual file can be usefull during development (remember that this file currently isn't really useful for the "normal" end-user anyway, as very few libraries support CMake in the first place).

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+set(BOOST_MAIN_IGNORE_LIBS callable_traits;hof;compute;gil;hana;yap;safe_numerics;beast CACHE string "List of libraries that will be excluded from cmake build")
+
+
+#~~~ options ~~~
+
+message("
+================================ BOOST WARNING ===================================================
+=                                                                                                =
+= General cmake support in boost is EXPERIMENTAL at best and                                     =
+= only a few libraries can be build via cmake at all.                                            =
```
Comment (orgads) [inline CMakeLists.txt]: built

File: libs/numeric/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2018 Mike Dev
```
Comment (orgads) [inline libs/numeric/CMakeLists.txt]: 2019

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Copyright 2018 Mike Dev
```
Comment (orgads) [inline CMakeLists.txt]: 2019

Comment (Mike-Devel) [reply CMakeLists.txt]: Well, the file was created in 2018. But as the have been modifications in 2019 I probably should add that year.

File: CMakeLists.txt (hunk 1 above)
Comment (Mike-Devel) [reply CMakeLists.txt]: Will fix

**Comment by Mike-Devel** [comment]
Superseded by pdimov's addition. Thanks!

**Pull request closed without merging**